File: csharp/src/Fib.cs
// The system under test: a tiny Fibonacci library.
//
// Every omni port ships this same library, with the same behaviour and the
// same error messages, and tests it with the same spec file
// (spec/fib.json).

using System;
using System.Collections.Generic;

namespace Voxgig.Omni.Test
{
    public static class Fib
    {
        /// <summary>Validate a Fibonacci index. The spec matches on these messages.</summary>
        public static long FibIndex(object val)
        {
            if (!Util.IsNum(val))
            {
                throw new ArgumentException("fib: not a number: " + Util.Stringify(val));
            }

            double num = Util.ToNum(val);

            if (num != Math.Truncate(num))
            {
                throw new ArgumentException("fib: non-integer index: " + Util.Stringify(val));
            }

            if (0 > num)
            {
                throw new ArgumentException("fib: negative index: " + Util.Stringify(val));
            }

            return (long)num;
        }

        /// <summary>The nth Fibonacci number: fib(0)=0, fib(1)=1.</summary>
        public static long FibNum(long index)
        {
            if (0 == index)
            {
                return 0;
            }

            long prev = 0;
            long cur = 1;

            for (long step = 1; step < index; step++)
            {
                long next = prev + cur;
                prev = cur;
                cur = next;
            }

            return cur;
        }

        public static object FibOf(object n) => FibNum(FibIndex(n));

        /// <summary>The first n Fibonacci numbers.</summary>
        public static object FibSeq(object n)
        {
            long count = FibIndex(n);
            var outlist = new List<object>();
            for (long index = 0; index < count; index++)
            {
                outlist.Add(FibNum(index));
            }
            return outlist;
        }

        /// <summary>The Fibonacci numbers from index start to index end, inclusive.</summary>
        public static object FibRange(object start, object end)
        {
            long from = FibIndex(start);
            long to = FibIndex(end);
            var outlist = new List<object>();
            for (long index = from; index <= to; index++)
            {
                outlist.Add(FibNum(index));
            }
            return outlist;
        }

        /// <summary>
        /// A description of the nth Fibonacci number. `prev` is null at index 0,
        /// which exercises the runner's null handling.
        /// </summary>
        public static object FibInfo(object n)
        {
            long index = FibIndex(n);
            long val = FibNum(index);

            return new Dictionary<string, object>
            {
                ["n"] = index,
                ["val"] = val,
                ["even"] = 0 == val % 2,
                ["prev"] = 0 == index ? null : (object)FibNum(index - 1),
                ["label"] = "fib(" + Util.NumStr(index) + ")=" + Util.NumStr(val),
            };
        }
    }
}

File: elixir/lib/fib.ex
# The system under test: a tiny Fibonacci library.
#
# Every omni port ships this same library, with the same behaviour and the
# same error messages, and tests it with the same spec file
# (spec/fib.json).

defmodule Voxgig.Omni.FibError do
  @moduledoc """
  The Fibonacci library's own error type. A subject must never raise the
  runner's OmniError - that type is reserved for test failures.
  """
  defexception [:message]
end

defmodule Voxgig.Omni.Fib do
  @moduledoc "A tiny Fibonacci library: the system under test."

  alias Voxgig.Omni.FibError
  alias Voxgig.Omni.Util, as: U

  @doc "Validate a Fibonacci index. The spec matches on these messages."
  def fibindex(val) do
    if not U.isnum(val) do
      raise FibError, message: "fib: not a number: " <> U.stringify(val)
    end

    num = val / 1

    if num != Float.round(num) do
      raise FibError, message: "fib: non-integer index: " <> U.stringify(val)
    end

    if 0 > num do
      raise FibError, message: "fib: negative index: " <> U.stringify(val)
    end

    trunc(num)
  end

  @doc "The nth Fibonacci number: fib(0)=0, fib(1)=1."
  def fibnum(0), do: 0
  def fibnum(index), do: fibstep(1, index, 0, 1)

  defp fibstep(step, index, _prev, cur) when step >= index, do: cur
  defp fibstep(step, index, prev, cur), do: fibstep(step + 1, index, cur, prev + cur)

  def fib(val), do: fibnum(fibindex(val)) / 1

  @doc "The first n Fibonacci numbers."
  def fibseq(val) do
    count = fibindex(val)
    Enum.map(0..(count - 1)//1, fn index -> fibnum(index) / 1 end)
  end

  @doc "The Fibonacci numbers from index start to index end, inclusive."
  def fibrange(start, stop) do
    from = fibindex(start)
    to = fibindex(stop)
    Enum.map(from..to//1, fn index -> fibnum(index) / 1 end)
  end

  @doc """
  A description of the nth Fibonacci number. `prev` is nil at index 0,
  which exercises the runner's null handling.
  """
  def fibinfo(val) do
    index = fibindex(val)
    num = fibnum(index)

    %{
      "n" => index / 1,
      "val" => num / 1,
      "even" => 0 == rem(num, 2),
      "prev" => if(0 == index, do: nil, else: fibnum(index - 1) / 1),
      "label" => "fib(" <> U.numstr(index) <> ")=" <> U.numstr(num)
    }
  end
end
